### ISSUE TITLE
OPTIONAL clauses include a redundant result if the value exists

### ISSUE THREAD

**Issue by dwhitney** [Callidon/sparql-engine#18]
**Describe the bug**
In an OPTIONAL clause if the result exists, an extra result will be produced containing a non-existent value. 

**To Reproduce**

The following query:

```
SELECT * 
WHERE {
  OPTIONAL {
    VALUES (?s ?p ?o) { ("s" "p" "o") }
  }
}
```
produces
```
[ 
  { '?s': '"s"', '?p': '"p"', '?o': '"o"' }, 
  {} 
]
```
**Expected behavior**

[ { '?s': '"s"', '?p': '"p"', '?o': '"o"' } ]

**Additional context**

I've written a test case: https://github.com/dwhitney/sparql-engine/blob/optional-problem/tests/sparql/optional-test.js#L37

I will try and resolve it this weekend.

**Comment by Callidon** [comment]
According to the SPARQL semantic [1], this is not a well-formed SPARQL query: an OPTIONAL clause is a left-join between two SPARQL group patterns. Here, you only have one group pattern (the VALUES clause). So, semantically speaking, the expected results for this query are not defined and anything is valid.

On the other hand, this might highlight a small bug with how the VALUES clause is handled. I've wrote the code for executing VALUES almost a year ago, so it could contains some surprise ^^

[1] Pérez, J., Arenas, M., & Gutierrez, C. (2009). [Semantics and complexity of SPARQL](https://users.dcc.uchile.cl/~cgutierr/papers/sparql.pdf). ACM Transactions on Database Systems (TODS), 34(3), 16.

**Comment by dwhitney** [comment]
Perhaps the SPARQL query is invalid, but the bug is real. I merely used the
VALUES clause as a way to highlight the bug without the need to add and
parse a new dataset. As described, when the value exists, you get a row
with the value present and a row without the value. If you have multiple
OPTIONAL clauses where this is the case, it's pretty bad because the result
set is the Cartesian product of all these rows. I will update the test case
and attempt a fix.

D

**Comment by dwhitney** [comment]
I've updated the test case

So with this data from section 6.2 of the SPARQL Spec

```
@prefix dc:   <http://purl.org/dc/elements/1.1/> .
@prefix :     <http://example.org/book/> .
@prefix ns:   <http://example.org/ns#> .

:book1  dc:title  "SPARQL Tutorial" .
:book1  ns:price  42 .
:book2  dc:title  "The Semantic Web" .
:book2  ns:price  23 .
```

with this query

```
    PREFIX  dc:  <http://purl.org/dc/elements/1.1/>
    PREFIX  ns:  <http://example.org/ns#>
    SELECT  ?title ?price
    WHERE   { 
      ?x dc:title ?title .
      OPTIONAL { 
        ?x ns:price ?price 
      }
    }
```

you'd expect two rows to be returned with the title and price of each book, but you get an extra third row as the test case show when it prints to the console:

```
[ { '?title': '"SPARQL Tutorial"',
    '?price': '"42"^^http://www.w3.org/2001/XMLSchema#integer' },
  { '?title': '"The Semantic Web"',
    '?price': '"23"^^http://www.w3.org/2001/XMLSchema#integer' },
  { '?title': '"SPARQL Tutorial"', '?price': 'UNBOUND' } ]
```

I'll see if I can figure it out

**Comment by Callidon** [comment]
Resolved in #19
